add localization support

L1cardo opened the pull request:
<img width="1436" alt="image" src="https://github.com/ExpressLRS/ExpressLRS-Configurator/assets/33802186/d1ef4bed-fef7-4b2c-9cb0-f0bd1ee72bb6">
<img width="1436" alt="image" src="https://github.com/ExpressLRS/ExpressLRS-Configurator/assets/33802186/283d6f99-0377-433c-8ce2-273578a6a5c7">

justinlampley commented:
Overall it looks very good!

I'll look at doing a more comprehensive review later, but there are a few things I noticed after a quick look through the changes.

- There are a few missing entries from the translations file, like ConfiguratorView.SelectFirmwareVersionFirst and SerialDeviceSelect.ManualSelection.  There may be more, but those are the ones I immediately noticed.

- For more complex text that includes react components, look at using the Trans component provided by react-i18next (https://react.i18next.com/latest/trans-component).  An example of where that component would be useful is the error message in the BuildResponse component.  Instead of breaking out the text into multiple translation entries and interweaving those between the link components, you can embed the components in the text which gives translators the flexibility of where the link components appear within the text for a particular language.  As an example, here is the build response error alert using the Trans component.

```
  <Alert sx={styles.errorMessage} severity="error">
	<AlertTitle>
	  {toTitle(
		response?.errorType ?? BuildFirmwareErrorType.GenericError
	  )}
	</AlertTitle>
	<p>
	  <Trans
		i18nKey="BuildResponse.ErrorMessage"
		defaults="An error has occured, see the above log for the exact error message. If you have not already done so, visit <ExpressLRSLink>Expresslrs.org</ExpressLRSLink> and read the <FlashingGuideLink>Flashing Guide</FlashingGuideLink> for your particular device as well as the <TroubleshootingGuideLink>Troubleshooting Guide</TroubleshootingGuideLink>.  If you are still having issues after reviewing the documentation, please copy the build logs above to an online paste site and post in the #help-and-support channel on the <DiscordLink>ExpressLRS Discord</DiscordLink> with a link to the logs and other relevant information like your device, which flashing method you were using, and what steps you have already taken to resolve the issue."
		components={{
		  ExpressLRSLink: (
			<DocumentationLink url="https://www.expresslrs.org/" />
		  ),
		  FlashingGuideLink: (
			<DocumentationLink url="https://www.expresslrs.org/quick-start/getting-started/" />
		  ),
		  TroubleshootingGuideLink: (
			<DocumentationLink url="https://www.expresslrs.org/quick-start/troubleshooting/#flashingupdating" />
		  ),
		  DiscordLink: (
			<DocumentationLink url="https://discord.gg/dS6ReFY" />
		  ),
		}}
	  />
	</p>
  </Alert>
 ```

- yarn.lock was changed a lot and the registry was changed from yarnpkg.com to npmjs.org, the only changes that should have occurred in this file is the addition of the new packages to support react-i18next.  It is quite strange that running yarn install to add the new packages would have changed the lock file that much.

pkendall64 commented on src/ui/components/UserDefinesAdvisor/index.tsx:
```diff
@@ -33,9 +35,7 @@ const UserDefinesAdvisor: FunctionComponent<UserDefinesAdvisorProps> = ({
-      messages.push(
-        'Disabling UART_INVERTED is uncommon. Please make sure that your transmitter supports that.'
-      );
+      messages.push(t('UserDefinesAdvisor.DisablUARTInvertedrWarning'));
```
DisableUARTInvertedWarning

pkendall64 commented on src/ui/components/WifiDeviceSelect/index.tsx:
```diff
@@ -122,13 +124,19 @@ const WifiDeviceSelect: FunctionComponent<WifiDeviceSelectProps> = (props) => {
-        <Tab label="WiFi Devices" value={WifiSourceType.LIST} />
-        <Tab label="Manual" value={WifiSourceType.MANUAL} />
+        <Tab
+          label={t('WifiDeviceSelect.WiFiDevices')}
```
Consistency, Wifi seems to be the thing used everywhere. i.e. lowercase `f`

pkendall64 commented on src/ui/views/ConfiguratorView/index.tsx:
```diff
@@ -1047,7 +1057,7 @@ const ConfiguratorView: FunctionComponent<ConfiguratorViewProps> = (props) => {
                   <ShowTimeoutAlerts
                     severity="warning"
-                    messages="Sometimes builds take at least a few minutes. It is normal, especially for the first time builds."
+                    messages={t('ConfiguratorView.BuildsTakeMminutesFirstTime')}
```
BuildsTakeMinutesFirstTime, extra 'm'

pkendall64 commented on src/ui/components/Header/index.tsx:
```diff
@@ -61,7 +64,7 @@ const Header: FunctionComponent = memo(() => {
       <Toolbar sx={styles.toolbar}>
         <Box sx={styles.logotype}>
-          <img src={LogotypeIcon} alt="ExpressLrs Configurator" />
+          <img src={LogotypeIcon} alt={t('Header.ExpressLrsConfigurator')} />
```
ExpressLRS (capitalisation)

L1cardo commented:
1. add missing localization entries
2. fix typos
3. improve complex text that includes react components

justinlampley commented on src/ui/components/UserDefineDescription/index.tsx:
```diff
@@ -33,351 +35,290 @@ const UserDefineDescription: FunctionComponent<UserDefineDescriptionProps> =
-                is 60 seconds.
-              </p>
+              <Trans i18nKey="UserDefineDescription.AutoWifiOnInterval">
+                <p>
```
Similar to the comment about the default property, no need to include the actual text when using the i18nKey property.

L1cardo replied:
Fixed.

L1cardo commented on src/ui/components/BuildResponse/index.tsx:
```diff
@@ -60,40 +63,32 @@ const BuildResponse: FunctionComponent<BuildResponseProps> = memo(
-              have already taken to resolve the issue.
+              <Trans
+                i18nKey="BuildResponse.ErrorDetails"
+                defaults="An error has occured, see the above log for the exact error message. If you have not 
```
Yeah, I know. I thought it would be friendly for developers if we leave the actual text there.
I will remove it if it is unnecessary.

justinlampley replied:
While I do see the thought behind it being helpful, the text in the Trans component can be complex, but since most of the translations are using the useTranslation hook and only supplying the key and not the text itself, only having the key for the Trans component would keep things consistent.  Also, since that text isn't actually used, it is just there for reference, it is likely to get out of sync with the translation file if is forgotten that both need to be updated.  At that point, the additional text that was supposed to be helpful can actually cause confusion since it says one thing, but what is actually appearing on the screen is something different.

L1cardo replied:
Fixed

L1cardo commented:
1. Removed default text valuefor <Trans> component
2. Fixed a typo in translation file en.json
3. Improved one zh-CN localization.

L1cardo commented on src/i18n/en/translation.json:
```diff
@@ -0,0 +1,213 @@
+    "FirmwareVersion": "Firmware version",
+    "Target": "Target",
+    "Notice": "Notice",
+    "SelectFirmwareVersionFirst.": "Please select a firmware version first.",
```
Fixed

jurgelenas commented on src/i18n/i18n.ts:
```diff
@@ -0,0 +1,20 @@
+import i18n from 'i18next';
+import { initReactI18next } from 'react-i18next';
+import LanguageDetector from 'i18next-browser-languagedetector'; // For auto-detecting language
+import Bankend from 'i18next-http-backend'; // For loading translations from backend
```
Should be `Backend`

jurgelenas commented on src/i18n/i18n.ts:
```diff
@@ -0,0 +1,20 @@
+  .use(initReactI18next)
+  .init({
+    fallbackLng: 'en',
+    debug: true,
```
`debug: true` - do we want it to be set to `true` in production builds too?

jurgelenas commented on src/i18n/i18n.ts:
```diff
@@ -0,0 +1,20 @@
+import i18n from 'i18next';
```
If we import this file as `src/i18n/i18n` in other modules I think it should be renamed to `src/i18n/index.ts`.

jurgelenas commented on src/index.tsx:
```diff
@@ -1,6 +1,7 @@
 import React from 'react';
 import { createRoot } from 'react-dom/client';
 import App from './ui/App';
+import './i18n/i18n';
```
After `src/i18n/i18n.ts` is renamed to `src/i18n/index.ts` it should be just `import './i18n'`

jurgelenas commented on src/ui/components/BuildResponse/index.tsx:
```diff
@@ -23,34 +24,36 @@ interface BuildResponseProps {
   ({ response, firmwareVersionData }) => {
+    const { t } = useTranslation();
+
     // TODO: translations
```
Remove `// TODO: translations`

jurgelenas commented on src/ui/views/SettingsView/index.tsx:
```diff
@@ -1,10 +1,12 @@
 import CardTitle from '../../components/CardTitle';
 import MainLayout from '../../layouts/MainLayout';
 
 const SettingsView: FunctionComponent = () => {
```
I think we should make the Settings page visible and allow our users to override the current language of the ExpressLRS Configurator.

jurgelenas commented:
@L1cardo can you please look into making translation type-safe according to https://www.i18next.com/overview/typescript ?

jurgelenas commented:
I have created a project on `crowdin.com`: https://crowdin.com/project/expresslrs-configurator

Will work on this: https://support.crowdin.com/github-integration/

L1cardo commented:
1. rename `i18n.ts` to `index.ts` 
2. remove `TODO: translations`
3. i18next debug true when only 'debug'
4. add settings view and a simple language switcher(I am really unfamiliar with ts coding, so this was very difficult for me, I wanted to add a drop down menu but I failed, so maybe really "programmers" can work on this)
<img width="400" alt="image" src="https://github.com/ExpressLRS/ExpressLRS-Configurator/assets/33802186/b5b08057-29ce-431a-b202-13b0fb2644c6">
<img width="400" alt="image" src="https://github.com/ExpressLRS/ExpressLRS-Configurator/assets/33802186/d9beec33-d8df-42c4-b580-5604673770cc">

L1cardo commented:
> I have created a project on `crowdin.com`: [crowdin.com/project/expresslrs-configurator](https://crowdin.com/project/expresslrs-configurator)
> 
> Will work on this: [support.crowdin.com/github-integration](https://support.crowdin.com/github-integration/)

For github and crowding integration, crowdin official integration only sync at the first time and we should manually sync every time after. https://support.crowdin.com/github-integration/#uploading-translations-from-repo.
For auto sync, you can take a look at how BetaFlight does.
https://github.com/betaflight/betaflight-configurator/blob/master/.github/workflows/translations-pr.yml
https://github.com/betaflight/betaflight-configurator/blob/master/.github/workflows/translations-upload.yml

L1cardo commented:
make ELRS Configurator title localizable
<img width="1436" alt="image" src="https://github.com/ExpressLRS/ExpressLRS-Configurator/assets/33802186/94f012f1-8325-4220-853c-241f6d8d11df">

L1cardo commented:
Before: 
<img width="176" alt="image" src="https://github.com/ExpressLRS/ExpressLRS-Configurator/assets/33802186/6fc2018e-321f-4042-9cad-f05e3abcf90c">

After: 
<img width="176" alt="image" src="https://github.com/ExpressLRS/ExpressLRS-Configurator/assets/33802186/1fc69af9-b6e1-4c24-b90e-b15366721930">

jurgelenas commented:
@L1cardo can you resolve merge conflicts?

jurgelenas commented on src/ui/views/SettingsView/index.tsx:
```diff
@@ -1,20 +1,43 @@
         <CardContent>
-          <Typography variant="h6">Todo:</Typography>
+          <Typography variant="h6">{t('SettingsView.LanguageSelector')}</Typography>
+          <ul>
```
What do you think about changing the list of buttons to our [<Omnibox/>](https://github.com/ExpressLRS/ExpressLRS-Configurator/tree/master/src/ui/components/Omnibox) component? Currently it looks fine, but we have only 2 languages at the moment. If we get 10 of them or more the current implementation will be not as usable.

jurgelenas commented on src/@types/i18next.d.ts:
```diff
@@ -0,0 +1,17 @@
+// import the original type declarations
```
Does this implementation work out of the box? Please double check if it is required to add `src/@types` path to the `tsconfig.json`.

L1cardo replied:
added `include src` to `tsconfig.json`

L1cardo commented:
1. resolve conflicts and update to latest code base
2. added mising localizations

L1cardo commented:
1. added language switcher and system info
2. added `src/@types` and `node_modules/@types`
<img width="400" alt="image" src="https://github.com/ExpressLRS/ExpressLRS-Configurator/assets/33802186/3bed7454-0081-4e31-b034-cb7d2615232c">
<img width="400" alt="image" src="https://github.com/ExpressLRS/ExpressLRS-Configurator/assets/33802186/24007001-0b92-4d56-bea8-88cf4fe9dbdd">

L1cardo commented:
I forgot to add `i18n` to resources or distribution app won't have i18n files, really sorry for that.
Here is the solution:
1. https://github.com/ExpressLRS/ExpressLRS-Configurator/blob/f6ccf9afe4992cc5276097e250be25c4116d6ddf/src/i18n/index.ts#L13
This should be `loadPath: process.env.NODE_ENV === 'production' ? '../src/i18n/{{lng}}/translation.json' : './i18n/{{lng}}/translation.json',`
<img width="1053" alt="image" src="https://github.com/ExpressLRS/ExpressLRS-Configurator/assets/33802186/11704448-f6b9-429e-b9c1-7250f36c5957">

2. https://github.com/ExpressLRS/ExpressLRS-Configurator/blob/f6ccf9afe4992cc5276097e250be25c4116d6ddf/package.json#L145
Should add `"./src/i18n/**"` to `extraResources`

<img width="237" alt="image" src="https://github.com/ExpressLRS/ExpressLRS-Configurator/assets/33802186/9eb337d0-472e-44c3-8f44-a04ea6c0fd0b">
